grpc/src/main/java/com/linecorp/armeria/server/grpc/UnaryServerCall.java

```diff
@@ -0,0 +1,215 @@
+import com.linecorp.armeria.common.grpc.protocol.DeframedMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.internal.common.grpc.GrpcLogUtil;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.grpc.CompressorRegistry;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+
+/**
+ * Encapsulates the state of a single server call, reading messages from the client, passing to business logic
+ * via {@link Listener}, and writing messages passed back to the response.
+ */
+final class UnaryServerCall<I, O> extends AbstractServerCall<I, O> {
+
+    private static final Logger logger = LoggerFactory.getLogger(UnaryServerCall.class);
+
+    private final HttpRequest req;
+    private final CompletableFuture<HttpResponse> resFuture;
+    private final ServiceRequestContext ctx;
+    private final UnaryMessageDeframer requestDeframer;
+
+    // Only set once.
+    @Nullable
+    private O responseMessage;
+    private boolean deframingStarted;
+
+    UnaryServerCall(HttpRequest req, MethodDescriptor<I, O> method, String simpleMethodName,
+                    CompressorRegistry compressorRegistry, DecompressorRegistry decompressorRegistry,
+                    HttpResponse res, CompletableFuture<HttpResponse> resFuture,
+                    int maxRequestMessageLength, int maxResponseMessageLength,
+                    ServiceRequestContext ctx, SerializationFormat serializationFormat,
+                    @Nullable GrpcJsonMarshaller jsonMarshaller, boolean unsafeWrapRequestBuffers,
+                    boolean useBlockingTaskExecutor, ResponseHeaders defaultHeaders,
+                    @Nullable GrpcStatusFunction statusFunction, boolean autoCompress) {
+        super(req, method, simpleMethodName, compressorRegistry, decompressorRegistry, res,
+              maxResponseMessageLength, ctx, serializationFormat, jsonMarshaller, unsafeWrapRequestBuffers,
+              useBlockingTaskExecutor, defaultHeaders, statusFunction, autoCompress);
+        requireNonNull(req, "req");
+        this.ctx = requireNonNull(ctx, "ctx");
+        final boolean grpcWebText = GrpcSerializationFormats.isGrpcWebText(serializationFormat);
+        requireNonNull(decompressorRegistry, "decompressorRegistry");
+
+        final RequestHeaders clientHeaders = req.headers();
+        requestDeframer = new UnaryMessageDeframer(ctx.alloc(), maxRequestMessageLength, grpcWebText)
+                .decompressor(clientDecompressor(clientHeaders, decompressorRegistry));
+        this.req = req;
+        this.resFuture = requireNonNull(resFuture, "resFuture");
+    }
+
+    @Override
+    public void request(int numMessages) {
+        if (ctx.eventLoop().inEventLoop()) {
+            request0();
+        } else {
+            ctx.eventLoop().execute(this::request0);
+        }
+    }
+
+    private void request0() {
```

Comment:
I will go with the latter one.

Reply:
On second thought, renaming looks better because `startDeframing()` is explicitly called when a `listener` is set for which nullness of `listener` need to check.